### PR TITLE
Implement new mapping on metal

### PR DESCRIPTION
**Pull request by omni-viral**

### CODE REVIEW

File: src/backend/metal/src/device.rs [hunk 1]
```diff
@@ -800,28 +800,64 @@ impl hal::Device<Backend> for Device {
...
     where
         I: IntoIterator,
         I::Item: Borrow<(&'a n::Memory, Range<u64>)>,
     {
-        unimplemented!()
+        for item in iter.into_iter() {
+            let (memory, ref range) = *item.borrow();
+            let mut mapping = memory.mapping.lock().unwrap();
+            assert!(mapping.is_some());
+            let mapping = mapping.as_ref().unwrap();
```
Comment (kvark) [inline src/backend/metal/src/device.rs]: the assert above is covered by this `unwrap`

Comment (omni-viral) [reply src/backend/metal/src/device.rs]: Correct.

File: src/backend/metal/src/native.rs [hunk 2]
```diff
@@ -226,7 +232,64 @@ pub enum DescriptorSetBinding {
 }
 
 #[derive(Debug)]
-pub enum Memory {
+pub struct Memory {
+    pub(crate) heap: MemoryHeap,
+    pub(crate) allocations: Arc<Mutex<MemoryAllocations>>,
+    pub(crate) mapping: Mutex<Option<MemoryMapping>>,
```
Comment (msiglreith) [inline src/backend/metal/src/native.rs]: A mutex is not strictly necessary here I think as `memory` in `vkMapMemory` and `vkUnmapMemory` should be externally synch following Vulkan.

File: src/backend/metal/src/native.rs [hunk 3]
```diff
@@ -226,7 +232,64 @@ pub enum DescriptorSetBinding {
...
+        buffers.extend(self.starts.range(range.clone()).map(|(&start, &(end, ref b))| (start .. end, b.clone())));
+        let range = (Bound::Excluded(range.start), Bound::Included(range.end));
+        buffers.extend(self.ends.range(range).map(|(&end, &(start, ref b))| (start .. end, b.clone())));
+        buffers.sort_unstable_by_key(|&(_, ref b)| b.as_ptr());
+        buffers.dedup_by_key(|&mut (_, ref b)| b.as_ptr());
+        buffers
+    }
+
+    pub(crate) fn insert(&mut self, range: Range<u64>, buffer: metal::Buffer) {
+        self.starts.insert(range.start, (range.end, buffer.clone()));
```
Comment (msiglreith) [inline src/backend/metal/src/native.rs]: Depending on the cost of cloning that buffer an additional `Vec<metal::Buffer>` might be handy and only storing indices in the trees.

Comment (omni-viral) [reply src/backend/metal/src/native.rs]: Cloning cost for buffer should be similar to `Arc`

File: src/backend/metal/src/native.rs (hunk 2 above)
Comment (omni-viral) [reply src/backend/metal/src/native.rs]: `UnsafeCell` then?

Comment (kvark) [reply src/backend/metal/src/native.rs]: Actually, this is not a perf concern. We can leave a `//TODO` comment and proceed.

**Comment by kvark** [comment]
Bors r+

**Comment by kvark** [comment]
Bors r+

**Pull request merged**